pulsar/consumer_partition.go

```diff
@@ -971,9 +972,9 @@ func (pc *partitionConsumer) grabConn() error {
 		pc.name = res.Response.ConsumerStatsResponse.GetConsumerName()
 	}
 
-	pc.conn = res.Cnx
+	pc._setConn(res.Cnx)
 	pc.log.Info("Connected consumer")
-	pc.conn.AddConsumeHandler(pc.consumerID, pc)
+	pc._getConn().AddConsumeHandler(pc.consumerID, pc)
```

Comment:
What's the behavior if _getConn() ever returns nil? Same for all the other places.

Reply:
@cckellogg ,

Good question; `_getConn()` should never return `nil`.

An invariant in this code is that the `partitionConsumer.conn` field must be set and is never `nil`.
The `grabConn()` method sets the `partitionConsumer.conn` field; `grabConn()` is called from the `newPartitionConsumer` factory method which will fail construction of the partitionConsumer if grabConn() returns an error.

The above said, it is probably better for the cast in `_getConn()` to be unchecked and let the code `panic()` if the invariant is broken.  Thoughts?

Reply:
I've made the cast unchecked and added a comment that explains why.

Reply:
There was change that moved the broker reconnect out of the events go routine (https://github.com/apache/pulsar-client-go/pull/376/files). This is now causing the data race issue. 

The question is if there is a reconnection what should happen with the pending events in the event channel? Right now they are processed using a stale/closed connection. Even with this fix it's possible events will try to get processed using a stale connection. Maybe that is ok but for me it makes the code difficult to follow and reason about. Ideally, we could come up with a cleaner way so we are never unintentionally using a stale/closed connection.

Thoughts?

Reply:
@cckellogg ,
I'm thinking that since [PR 376](https://github.com/apache/pulsar-client-go/pull/376) drains the connection incomingRequestsCh on close it should be possible remove the extra go-routine in the partitionConsumer and select from the connectionClosedCh in the partitonConsumer.runEventsLoop.  What do you think?

Reply:
@cckellogg ,
I've attempted the above with the following commit:
https://github.com/apache/pulsar-client-go/pull/535/commits/17335cd7846b71bd14592c944ea76f3fbcfdafd6

If this change is accepted I'll clean up this branch and update the PR description.

Reply:
In https://github.com/apache/pulsar-client-go/pull/535/commits/5dfcc1300eee30824efa5945bf463beb3bc3474e we also short-circuit broker re-connection attempts on consumer close.

Reply:
@cckellogg ,

I've reverted the behaviour of this PR back to the first approach, making access to the connection in the partitionConsumer atomic.

Reply:
The latest CI failure is going to be addressed by https://github.com/apache/pulsar-client-go/pull/544